backwardcompatibilityml/loss/strict_imitation.py

```diff
@@ -133,7 +134,145 @@ def forward(self, x, y, reduction="mean"):
             h2_support_output_logit, _, _ = self.h2(x_support)
             dissonance = self.dissonance(h1_support_output_labels, h2_support_output_logit)
 
-        base_loss = self.loss(h2_output_logit, y)
         base_loss = self.loss(h2_output_logit, y, reduction=reduction)
-        new_error_loss = base_loss + self.lambda_c * dissonance
-        return new_error_loss
+        strict_imitation_loss = base_loss + self.lambda_c * dissonance
+        return strict_imitation_loss
+
+
+class StrictImitationBinaryCrossEntropyLoss(nn.Module):
+    """
+    Strict Imitation Cross Entropy Loss
+
+    This class implements the strict imitation loss function
+    with the underlying loss function being the Cross Entropy loss.
+
+    Example usage:
+        h1 = MyModel()
+        ... train h1 ...
+        h1.eval() (it is important that h1 be put in evaluation mode)
+
+        lambda_c = 0.5 (regularization parameter)
+        h2 = MyNewModel() (this may be the same model type as MyModel)
+        siloss = StrictImitationBinaryCrossEntropyLoss(h1, h2, lambda_c)
+
+        for x, y in training_data:
+            loss = siloss(x, y)
+            loss.backward()
+
+        Note that we pass in the input and the target directly to the
+        siloss function instance. It calculates the outputs of h1 and h2
+        internally.
+
+    Args:
+        h1: Our reference model which we would like to be compatible with.
+        h2: Our new model which will be the updated model.
+        lambda_c: A float between 0.0 and 1.0, which is a regularization
+            parameter that determines how much we want to penalize model h2
+            for being incompatible with h1. Lower values panalize less and
+            higher values penalize more.
+    """
+
+    def __init__(self, h1, h2, lambda_c, discriminant_pivot=0.5, **kwargs):
+        super(StrictImitationBinaryCrossEntropyLoss, self).__init__()
+        self.h1 = h1
+        self.h2 = h2
+        self.loss = F.binary_cross_entropy
+        self.lambda_c = lambda_c
+        self.discriminant_pivot = discriminant_pivot
+
+    def dissonance(self, h1_output_sigmoid, h2_output_sigmoid):
```

Comment:
@nushib If you take a look at how the dissonance is being calculated for new error vs strict imitation. The end result is the same, i.e. we are calculating how h2 differs from the target labeling in either case.

This makes me think that I have interpreted the strict imitation loss incorrectly, at least in the case of Binary Cross Entropy.

Should I instead just be measuring the negative_log_likelihood(h2, h1) in the cases where h1 is correct? For each and every strict imitation dissonance?

Or should i be measuring parent_loss(h2, h1) in cases where h1 is correct.

If the latter, then depending on the type of parent_loss, i.e. NLL, CrossEntropy, KLDiv or BinaryCrossEntropy ... how we define this is open to interpretation.

Reply:
Based on discussion offline. Will have the dissonance in the Strict Imitation Binary Cross Entropy loss be calculated using the negative log likelihood. And will make sure to document why we are doing this so that users are aware.